design/ecs-xray-support.md

```diff
@@ -0,0 +1,151 @@
+# AWS ECS - Support for Instrumenting ECS Applications with AWS X-Ray
+
+In order to provide an AWS native solution for instrumenting ECS applications, the ECS construct library should provide an API to integrate AWS-Ray into their Task Definition.
+
+Github Issue: https://github.com/awslabs/aws-cdk/issues/1843
+
+## General approach
+We would be adding an API to EC2Service and FargateService to inject X-ray as a sidecar container.
+
+*NOTE:* This assumes the customer has the necessary code changes in their application image to interact with AWS X-Ray. See the [AWS X-Ray developer Guide](https://docs.aws.amazon.com/xray/latest/devguide/xray-daemon-ecs.html) for more details.
+
+For a FargateService, where the network mode is awsvpc, it should suffice to wrap the `addContainer` call on the FargateTaskDefinition, and add the appropriate port mappings (x-ray by default listens on port 2000):
+
+```ts
+public addTracing(props: TracingOptions = {}) {
```

Comment:
Were we not thinking of adding this as a `Feature`? Do we still have that, or not?

Reply:
Can you remind me of what the Feature feature entails again?

Reply:
We would make something like this:

```ts
taskDefinition.addFeature(new XRayFeature({ ... }));
```

It would then be trivial to also allow this:

```ts
new TaskDefinition(...., {
  features: [new XRayFeature({ ... })]
});
```

But yeah. Not necessarily required.

Reply:
Would adding a feature to a TaskDef essentially just add a container with the properties in the feature?